File: vitral/src/vsdk/toolkit/common/Vertex.java
package vsdk.toolkit.common;

import java.util.ArrayList;

public class Vertex extends Entity
{
    /// Check the general attribute description in superclass Entity.
    public static final long serialVersionUID = 20060502L;

    ///
    public Vector3D position;

    ///
    public Vector3D normal;

    ///
    public Vector3D binormal;

    ///
    public Vector3D tangent;

    /// Texture coordinates
    public double u;
    public double v;

    ///
    public ArrayList<Triangle> incidentTriangles;

    public Vertex() {
        position = null;
        normal = null;
        binormal = null;
        tangent = null;
        this.incidentTriangles = null;
        u = 0.0;
        v = 0.0;
    }

    public Vertex(double x, double y, double z) {
        this.position = new Vector3D(x, y, z);
        normal = new Vector3D(1, 0, 0);
        binormal = new Vector3D(0, 1, 0);
        tangent = new Vector3D(0, 0, 1);
        incidentTriangles = null;
        u = 0.0;
        v = 0.0;
    }

    public Vertex(Vector3D position) {
        this.position = position;
        normal = new Vector3D(1, 0, 0);
        binormal = new Vector3D(0, 1, 0);
        tangent = new Vector3D(0, 0, 1);
        incidentTriangles = null;
        u = 0.0;
        v = 0.0;
    }

    public Vertex(Vector3D position, Vector3D normal) {
        this.position = position;
        this.normal = normal;
        this.normal.normalize();

        binormal = new Vector3D(0, 1, 0);
        tangent = new Vector3D(0, 0, 1);

        binormal = null;
        tangent = null;
        incidentTriangles = null;
        u = 0.0;
        v = 0.0;
    }

    public Vertex(Vector3D position, Vector3D normal, Vector3D binormal,
                  Vector3D tangent) {
        this.position = position;
        this.normal = normal;
        this.binormal = binormal;
        this.tangent = tangent;
        incidentTriangles = null;
        u = 0.0;
        v = 0.0;
    }

    public Vertex(Vertex vertex) {
        this.position = vertex.position;
        this.normal = vertex.normal;
        this.binormal = vertex.binormal;
        this.tangent = vertex.tangent;
        this.incidentTriangles=vertex.getIncidentTriangles();
        this.u = vertex.u;
        this.v = vertex.v;
    }

    public Vector3D getPosition() {
        return this.position;
    }

    public Vector3D getNormal() {
        return normal;
    }

    public Vector3D getBinormal() {
        return this.binormal;
    }

    public Vector3D getTangent() {
        return this.tangent;
    }

    public double getU() {
        return this.u;
    }

    public double getV() {
        return this.v;
    }

    public ArrayList<Triangle> getIncidentTriangles() {
        return this.incidentTriangles;
    }

    public void setPosition(Vector3D position) {
        this.position = new Vector3D(position);
    }

    public void setNormal(Vector3D normal) {
        this.normal = new Vector3D(normal);
        this.normal.normalize();
    }

    public void setBinormal(Vector3D binormal) {
        this.binormal = binormal;
    }

    public void setTangent(Vector3D tangent) {
        this.tangent = tangent;
    }

    public void setU(double u) {
        this.u = u;
    }

    public void setV(double v) {
        this.v = v;
    }

    public void setIncidentTriangles(ArrayList<Triangle> incidentTriangles) {
        this.incidentTriangles = incidentTriangles;
    }

    public Triangle getIncidentTriangleAt(int index) {
        return (Triangle) incidentTriangles.get(index);
    }

    /**
    Provides an object to text report convertion, optimized for human
    readability and debugging. Do not use for serialization or persistence
    purposes.
    */
    public String toString() {
        return "v < " + position.x + ", " + position.y + ", " +
            position.z + " > <"+u+", "+v+">";
    }
}

File: cpp/base/src/main/vsdk/toolkit/environment/geometry/elements/Vertex.cpp
#include "Vertex.h"
#include "Triangle.h"
#include "vsdk/toolkit/common/VSDK.h"
#include "java/util/ArrayList.h"
#include "java/util/ArrayList.txx"

Vertex::Vertex(double x, double y, double z)
    : position(x, y, z), normal(1, 0, 0), binormal(0, 1, 0), tangent(0, 0, 1),
      u(0.0), v(0.0), incidentTriangles(nullptr) {}

Vertex::Vertex(const Vector3Dd& position)
    : position(position), normal(1, 0, 0), binormal(0, 1, 0), tangent(0, 0, 1),
      u(0.0), v(0.0), incidentTriangles(nullptr) {}

Vertex::Vertex(const Vector3Dd& position, const Vector3Dd& normal)
    : position(position), normal(normal.normalized()), binormal(0, 1, 0), tangent(0, 0, 1),
      u(0.0), v(0.0), incidentTriangles(nullptr) {}

Vertex::Vertex(const Vector3Dd& position, const Vector3Dd& normal, double inU, double inV)
    : position(position), normal(normal.normalized()), binormal(0, 1, 0), tangent(0, 0, 1),
      u(inU), v(inV), incidentTriangles(nullptr) {}

Vertex::Vertex(const Vector3Dd& position, const Vector3Dd& normal, const Vector3Dd& binormal, const Vector3Dd& tangent)
    : position(position), normal(normal), binormal(binormal), tangent(tangent),
      u(0.0), v(0.0), incidentTriangles(nullptr) {}

Vertex::Vertex(const Vertex& vertex)
    : position(vertex.position), normal(vertex.normal), binormal(vertex.binormal), tangent(vertex.tangent),
      u(vertex.u), v(vertex.v), incidentTriangles(vertex.incidentTriangles) {}

Vector3Dd Vertex::getPosition() const { return position; }
Vector3Dd Vertex::getNormal() const { return normal; }
Vector3Dd Vertex::getBinormal() const { return binormal; }
Vector3Dd Vertex::getTangent() const { return tangent; }
double Vertex::getU() const { return u; }
double Vertex::getV() const { return v; }
java::ArrayList<Triangle>* Vertex::getIncidentTriangles() const { return incidentTriangles; }

void Vertex::setPosition(const Vector3Dd& inPosition) { position = Vector3Dd(inPosition); }
void Vertex::setNormal(const Vector3Dd& inNormal) { normal = Vector3Dd(inNormal).normalized(); }
void Vertex::setBinormal(const Vector3Dd& inBinormal) { binormal = inBinormal; }
void Vertex::setTangent(const Vector3Dd& inTangent) { tangent = inTangent; }
void Vertex::setU(double inU) { u = inU; }
void Vertex::setV(double inV) { v = inV; }
void Vertex::setIncidentTriangles(java::ArrayList<Triangle>* inIncidentTriangles) { incidentTriangles = inIncidentTriangles; }

Triangle Vertex::getIncidentTriangleAt(int index) const
{
    return incidentTriangles->get(index);
}

std::string Vertex::toString() const
{
    return "v <" + VSDK::formatDouble(position.x()) + ", " +
        VSDK::formatDouble(position.y()) + ", " +
        VSDK::formatDouble(position.z()) + "> n <" +
        VSDK::formatDouble(normal.x()) + ", " +
        VSDK::formatDouble(normal.y()) + ", " +
        VSDK::formatDouble(normal.z()) + "> UV<" +
        VSDK::formatDouble(u) + ", " + VSDK::formatDouble(v) + ">";
}
